### ISSUE TITLE
Add Top-nσ logit truncation custom logit processor example

### ISSUE THREAD

**Issue by Jackie2049** [sgl-project/sglang#29397]
## Motivation

SGLang's custom logit processor framework (enabled via `--enable-custom-logit-processor`) lets users inject arbitrary logit-space processing per-request. However, the only existing reference is a **dummy** `DeterministicLogitProcessor` in the docs that always samples a fixed token — there is no example of a processor that does meaningful logit filtering.

This issue proposes adding the **first real-world custom logit processor example**: [Top-nσ](https://arxiv.org/abs/2411.07641) ("Top-nσ: Not All Logits Are You Need", ACL 2025), a logit-space dynamic truncation method.

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike sglang's built-in probability-space filters (`top_k`, `top_p`, `min_p`), Top-nσ operates **before softmax** (where SGLang's custom processor pipeline runs) and adapts the candidate set to the "peakiness" of the distribution.

### How it compares to existing methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: sglang's top_k/top_p/min_p all operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space). Top-nσ is the first logit-space truncation reference for the custom processor framework.

### Why Top-nσ?

- **Truncation before softmax** — filters in logit space first, then softmax only processes the surviving candidates.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally).
- **Argmax-invariant** — the top-1 token always survives, so greedy decoding is unaffected.

## Proposal

Add an example in a new `examples/runtime/custom_logit_processor/` folder — `top_n_sigma.py` — following the existing `examples/runtime/` pattern (standalone script, documented server launch command, client-side usage).

**No framework changes**: the processor is defined in the example file and injected per-request via the existing `custom_logit_processor` field + `--enable-custom-logit-processor` gate.

### Implementation characteristics

- **Fully vectorized** across the batch (no per-row Python loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula (avoids a duplicate vocab-wide reduction)
- **Argmax-invariant** by construction

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

## References

Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

**Comment by ajinkyajawale14499** [comment]
@Jackie2049 Can I pick this one?

**Comment by Jackie2049** [comment]
@ajinkyajawale14499 I already have a draft PR up at #29398 with an initial implementation, but there's definitely room for improvement. Would you like to contribute together and improve the current version?

It's a single file (`examples/runtime/custom_logit_processor/top_n_sigma.py`), so easy to jump into. Here are some areas you could take ownership of:

- **End-to-end test**: Run it against a real model with `--enable-custom-logit-processor` to validate everything works end-to-end
- **OpenAI path polish**: Add more edge cases or variations in the OpenAI example
- **Code cleanup & performance optimization**: Polish code style, apply reviewer suggestions, and find opportunities to make the processor faster or handle more edge cases

Feel free to push commits directly to the PR branch or open suggestions. Happy to co-author or let you own specific parts — let's ship this together!
